### PR TITLE
[SPARK-42496][CONNECT][DOCS] Adding Spark Connect to the Spark 3.4 documentation

### PR DESCRIPTION
**Pull request by aldogrut**
### What changes were proposed in this pull request?
Adding a Spark Connect overview page to the Spark 3.4 documentation and a short section on the Spark overview page with a link to it.

### Why are the changes needed?
The first version of Spark Connect is released as part of Spark 3.4.0 and this adds an overview for it to the documentation.

### Does this PR introduce _any_ user-facing change?
Yes, the user facing documentation is updated.

### How was this patch tested?
Built the doc website locally and tested the pages.
SKIP_SCALADOC=1 SKIP_RDOC=1 bundle exec jekyll build

index.html
index html

spark-connect-overview.html
spark-connect-overview html

### CODE REVIEW

File: docs/spark-connect-overview.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+{% endhighlight %}
+
+# Client application authentication
+
+While Spark Connect does not have built-in authentication, it is designed to work seamlessly with your existing authentication infrastructure. Its gRPC HTTP/2 interface allows for the use of authenticating proxies, which makes it possible to secure Spark Connect without having to implement authentication logic in Spark directly.
+
+# What is included in Spark 3.4
+
+In Spark 3.4, Spark Connect provides DataFrame API coverage for PySpark and 
+Spark Connect clients for other languages are planned for the future.
```
Comment (itholic) [inline docs/spark-connect-overview.md:108]: I believe it is important to inform users in advance that while Spark Connect supports many of the key features of PySpark, there are also some features that are not supported.

For example, we can write something like this:

In Spark 3.4, Spark Connect supports most of the key APIs of PySpark, including DataFrame, Functions, and Column. However, some APIs such as SparkContext and RDD are not supported yet. The support for specific PySpark APIs in Spark Connect is indicated in all PySpark API references with phrases like “Support Spark Connect”. Therefore, it is recommended to check in advance whether the APIs you are using are supported by Spark Connect before migrating existing code. Support for Spark Connect clients in other languages is planned for the future.

the below is including the link as well just for your convenience (when you built the docs, the link would work):

```md
In Spark 3.4, Spark Connect supports most of the key APIs of PySpark, including [DataFrame](api/python/reference/pyspark.sql/dataframe.html), [Functions](api/python/reference/pyspark.sql/functions.html), and [Column](api/python/reference/pyspark.sql/column.html). However, some APIs such as [SparkContext](api/python/reference/api/pyspark.SparkContext.html#pyspark.SparkContext) and [RDD](api/python/reference/api/pyspark.RDD.html#pyspark.RDD) are not supported yet. The support for specific PySpark APIs in Spark Connect is indicated in all PySpark API references with phrases like "Support Spark Connect". Therefore, it is recommended to check in advance whether the APIs you are using are supported by Spark Connect before migrating existing code. Support for Spark Connect clients in other languages is planned for the future.
```

File: docs/spark-connect-overview.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+## Set SPARK_REMOTE environment variable
+
+If you set the SPARK_REMOTE environment variable on the client machine where your Spark client application is running and create a new Spark Session as illustrated below, the session will be a Spark Connect session. With this approach, there is no code change needed to start using Spark Connect.
+
+Set SPARK_REMOTE environment variable:
+
+{% highlight bash %}
+    export SPARK_REMOTE="sc://localhost/"
+{% endhighlight %}
```
[the remark this answers is not in the document]
Comment (allanf-db) [reply docs/spark-connect-overview.md:66]: Good point, fixing this

File: docs/index.md [hunk 3]
```diff
@@ -86,6 +88,15 @@ Example applications are also provided in R. For example,
 
     ./bin/spark-submit examples/src/main/r/dataframe.R
 
+## Running Spark Client Applications Anywhere with Spark Connect
```
[the remark this answers is not in the document]
Comment (allanf-db) [reply docs/index.md:91]: Fixing, thanks

File: docs/index.md [hunk 4]
```diff
@@ -49,8 +49,19 @@ For Java 11, `-Dio.netty.tryReflectionSetAccessible=true` is required additional
...
 
-Spark comes with several sample programs.  Scala, Java, Python and R examples are in the
-`examples/src/main` directory. To run one of the Java or Scala sample programs, use
+Spark comes with several sample programs. Python, Scala, Java and R examples are in the
+`examples/src/main` directory.
+
+To run Spark interactively in a Python interpreter, use
+`bin/pyspark`:
+
+    ./bin/pyspark --master local[2]
```
[the remark this answers is not in the document]
Comment (allanf-db) [reply docs/index.md:58]: Yeah, I did not update the existing code samples on this page (yet)

File: docs/spark-connect-overview.md (hunk 1 above)
Comment (allanf-db) [reply docs/spark-connect-overview.md]: I added this content, thanks

File: docs/index.md [hunk 5]
```diff
@@ -86,6 +88,16 @@ Example applications are also provided in R. For example,
...
 
+## Running Spark Client Applications Anywhere with Spark Connect
+
+Spark Connect is a new client-server architecture introduced in Spark 3.4 that decouples Spark
+client applications and allows remote connectivity to Spark clusters. The separation between
+client and server allows Spark and its open ecosystem to be leveraged from anywhere, embedded
+in any application. In Spark 3.4, Spark Connect provides DataFrame API coverage for PySpark.
+Spark Connect clients for other languages are planned for the future.
+
+To learn more about Spark Connect and how to use it, visit [Spark Connect Overview](spark-connect-overview.html).
```
Comment (itholic) [inline docs/index.md:99]: qq: It seems that the only way to access the Spark Connect Overview is through the link provided at this point.
Do you think that's enough, or maybe should we also place the link in a more prominent location to expose it more actively to the users?

**Review by itholic — APPROVED** [review]
Looks good otherwise!

File: docs/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (itholic) [reply docs/index.md:58]: Yeah, I think we can just update this to `./bin/pyspark --master local` or just simply `./bin/pyspark` while we're here :-)

Both way looks fine to me if it's working properly (I checked both work fine in my local env)

File: docs/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (allanf-db) [reply docs/index.md:91]: Fixed

File: docs/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (allanf-db) [reply docs/index.md:58]: I tried on my machine and it fails for me as well so I removed "--master local[2]" from these examples and then they work for me.

[the remark this answers is not in the document]
Comment (itholic) [reply docs/index.md]: Sounds good!

File: docs/spark-connect-overview.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+If you set the `SPARK_REMOTE` environment variable on the client machine where your
+Spark client application is running and create a new Spark Session as illustrated
+below, the session will be a Spark Connect session. With this approach, there is
+no code change needed to start using Spark Connect.
+
+Set the `SPARK_REMOTE` environment variable to point to the Spark server we started
+above:
+
+{% highlight bash %}
+export SPARK_REMOTE="sc://localhost/"
```
Comment (HyukjinKwon) [inline docs/spark-connect-overview.md:135]: nit but I would remove `/` in the end at `sc://localhost/"` -> `sc://localhost"`

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

File: docs/spark-connect-overview.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+>>> df = spark.createDataFrame(data).toDF(*columns)
+>>> df.show()
++---+-----+
+| id| name|
++---+-----+
+|  1|Sarah|
+|  2|Maria|
++---+-----+
+
+>>>
```
Comment (HyukjinKwon) [inline docs/spark-connect-overview.md:206]: nit but I would remove this :-). Please make a followup PR if you find some time.

**Pull request closed without merging**